### PR TITLE
crush: add module

### PR DESCRIPTION
**Pull request by digitalrane**
### Description

This adds a module for Charmbracelet Crush, an agentic coding tool.

The module supports:
- Setting main Crush configuration
- Configuring LSP servers
- Configuring MCP servers
- Configuring AI providers
- Configuring skills directory, or by setting a skills directory
- Including MCP server configuration via `programs.mcp`
- Secrets handling for sensitive values

This also enables non-free package installation in tests, which is required for some of the Nix build tests to pass in CI. Whilst not required for the unit tests included in this PR, other generic tests fail on this PR without the ability to install non-free software.

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `nix fmt` or
      `nix-shell -p treefmt nixfmt deadnix keep-sorted nixf-diagnose --run treefmt`.

- [X] Code tested through `nix run .#tests -- test-all` or
      `nix-shell --pure tests -A run.all`.

- [X] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [X] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [X] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [X] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [X] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

### CODE REVIEW

File: modules/programs/crush.nix [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+          };
+
+          # Final merged configuration
+          finalConfig = cfg.settings // mergedLsp // mergedMcp // mergedProviders;
+        in
+        {
+          ".config/crush/crush.json" = lib.mkIf (finalConfig != { } && finalConfig != null) {
+            source = jsonFormat.generate "crush-config.json" finalConfig;
+          };
+
```
Comment (khaneliman) [inline modules/programs/crush.nix]: Way too many options for stuff that amounts to a simple `json` generator.

Comment (digitalrane) [reply modules/programs/crush.nix]: @khaneliman are you willing to expand on this at all? Were there specific options that you didn't think were necessary? I've tried to be quite diligent in providing examples and validation of user options so users of the module know how things need to be structured, as well as providing tests for the various combinations of options, and an integration with the shared `programs.mcp` namespace, which in my mind was all to make the module more useful.

Comment (khaneliman) [reply modules/programs/crush.nix]: https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md essentially is the aim for modules. Prefer simple and flexible freeform modules with generators, as much as possible. The more complexity added with options, the larger the maintenance burden and chance you will block valid configs because of misunderstanding upstream or upstream deviations.

Comment (digitalrane) [reply modules/programs/crush.nix]: The RFC reference is useful, thanks - I'll refactor the module to follow this guidance.

Comment (digitalrane) [reply modules/programs/crush.nix]: @khaneliman I've removed all the extra options & simplified the option filtering/validation logic - I've kept the secrets handling wrapper & `programs.mcp` integration, and updated the tests (which are passing). I also added a couple of extra tests to do some further checking as the module isn't structuring the inputs as strictly now, so there was a couple of extra cases to cover. Please take a look and let me know if you've got further feedback. Thanks!

File: tests/modules/programs/crush/default.nix [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  crush-mcp-no-type = ./mcp-no-type.nix;
+  crush-mcp-shared-config = ./mcp-shared-config.nix;
+  crush-options-only = ./options-only.nix;
+  crush-providers-config = ./providers-config.nix;
+  crush-providers-false = ./providers-false.nix;
+  crush-providers-minimal = ./providers-minimal.nix;
+  crush-secret-with-settings = ./secret-with-settings.nix;
+  crush-secret-wrapper = ./secret-wrapper.nix;
+  crush-skills-config = ./skills-config.nix;
+  crush-skills-dir = ./skills-dir.nix;
```
Comment (khaneliman) [inline tests/modules/programs/crush/default.nix]: Could we please limit the number of test cases to be unique branches. We don't need to test the same exact thing in multiple files. 

ie: 

Full config would basically cover most testing. But, you might want to cover something that CANNOT be done within the same configuration in another test case.

Comment (digitalrane) [reply tests/modules/programs/crush/default.nix]: @khaneliman - I've consolidated all of the configuration tests into a single test & expected config, with the exception of the skills handling, which has two distinct modes, setting the directory to a non-managed location (for example, if you want to use the same skills repo and manage it separately), or managing the skills directory, contents or links to other files decoratively. These are distinct options to enable the two distinct workflows for using and managing available skills. I've also got another test for validating that an empty config will not produce a configuration file, which is desired behavior. I've also made some of the comments and documentation more concise. Let me know if you've got further feedback!

File: modules/programs/crush.nix [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        Main JSON configuration for Crush.
+        Writes to {file}`$HOME/.config/crush/crush.json`.
+
+        Configure LSP servers ({option}`settings.lsp`), MCP servers ({option}`settings.mcp`),
+        AI providers ({option}`settings.providers`), and general options ({option}`settings.options`).
+
+        For sensitive values like API keys, use {option}`secretEnvVars`
+        to inject them from files at runtime.
+
+        See <link xlink:href="https://github.com/charmbracelet/crush"/> for full configuration options.
```
Comment (rycee) [inline modules/programs/crush.nix]:
```suggestion
        See <https://github.com/charmbracelet/crush> for full configuration options.
```

Comment (digitalrane) [reply modules/programs/crush.nix]: Thanks @rycee - I've fixed that. I'll use that formatting in the future. Do you know if there is a good reference for the option description strings and how to mark them up for this project? The `link xref` part is used in nixpkgs and I picked it up from there - is it a project-specific preference?

File: all-maintainers.nix [hunk 4]
```diff
@@ -951,6 +951,14 @@
...
   };
+  rane = {
+    name = "Rane";
+    email = "rane+git@junkyard.systems";
+    matrix = "@rane:junkyard.systems";
+    github = "digitalrane";
+    githubId = 1829286;
+    keys = [ { fingerprint = "EBB6 0EE1 488F D04C D922  C039 AE96 1AF5 9D40 10B5"; } ];
+  };
   diniamo = {
```
Comment (khaneliman) [inline all-maintainers.nix:962]: separate first commit

File: flake.nix [hunk 5]
```diff
@@ -54,10 +54,18 @@
           "x86_64-linux"
         ];
 
+        # Helper to get pkgs with unfree allowed for tests
+        pkgsWithUnfree =
+          system:
+          import nixpkgs {
+            inherit system;
+            config.allowUnfree = true;
+          };
```
Comment (khaneliman) [inline flake.nix:63]:
```suggestion
        pkgsWithUnfree =
          system:
          import nixpkgs {
            inherit system;
            config.allowUnfree = true;
          };
```

separate commit

File: modules/programs/crush.nix [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+              withProviders =
+                if cfg.settings ? providers then
+                  withMcp // { providers = lib.mapAttrs (n: v: filterNulls v) cfg.settings.providers; }
+                else
+                  withMcp;
+            in
+            filterNulls withProviders;
+
+          finalConfig = filteredSettings;
+        in
```
Comment (khaneliman) [inline modules/programs/crush.nix:297]: Code smell of having so many if else conditions and nested. Would prefer more idiomatic nix. Also would prefer to reduce the amount of inline comments on basic things that the variable name will explain.

File: modules/programs/crush.nix [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+    if lib.isAttrs value then
+      let
+        filtered = lib.mapAttrs (n: v: filterNulls v) value;
+      in
+      lib.filterAttrs (n: v: v != null && v != { } && v != [ ]) filtered
+    else if lib.isList value then
+      map filterNulls value
+    else
+      value;
+
```
Comment (khaneliman) [inline modules/programs/crush.nix:23]: Doesn't feel like something that needs to exist. We should just be able to use nix module system for merging values and a generator for only operating on things that exist

**Comment by digitalrane** [comment]
I've decided to close this out, this was an experiment to see if this could be a useful tool for me, but I'm no longer using it. Thank you for the reviews & feedback and I'll keep them in mind for future contributions.
